### PR TITLE
meson: add unit tests

### PR DESCRIPTION
**Pull request by neheb**
Signed-off-by: Rosen Penev <rosenp@gmail.com>

### CODE REVIEW

**Comment by neheb** [comment]
```
 exiv2.lib(exiv2-1.dll) : error LNK2005: "public: __cdecl Exiv2::ExifData::~ExifData(void)" (??1ExifData@Exiv2@@QEAA@XZ) already defined in libexiv2int.a(src_tiffimage_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: unsigned __int64 __cdecl Exiv2::DataBuf::size(void)const " (?size@DataBuf@Exiv2@@QEBA_KXZ) already defined in libexiv2int.a(src_tiffimage_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: __cdecl Exiv2::DataBuf::~DataBuf(void)" (??1DataBuf@Exiv2@@QEAA@XZ) already defined in libexiv2int.a(src_tiffimage_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: virtual __cdecl Exiv2::Error::~Error(void)" (??1Error@Exiv2@@UEAA@XZ) already defined in libexiv2int.a(src_tiffimage_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: bool __cdecl Exiv2::DataBuf::empty(void)const " (?empty@DataBuf@Exiv2@@QEBA_NXZ) already defined in libexiv2int.a(src_pngchunk_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: __cdecl Exiv2::DataBuf::DataBuf(void)" (??0DataBuf@Exiv2@@QEAA@XZ) already defined in libexiv2int.a(src_pngchunk_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: struct Exiv2::DataBuf & __cdecl Exiv2::DataBuf::operator=(struct Exiv2::DataBuf &&)" (??4DataBuf@Exiv2@@QEAAAEAU01@$$QEAU01@@Z) already defined in libexiv2int.a(src_pngchunk_int.cpp.obj)
exiv2.lib(exiv2-1.dll) : error LNK2005: "public: class std::_Vector_iterator<class std::_Vector_val<struct std::_Simple_types<unsigned char> > > __cdecl Exiv2::DataBuf::begin(void)" (?begin@DataBuf@Exiv2@@QEAA?AV?$_Vector_iterator@V?$_Vector_val@U?$_Simple_types@E@std@@@std@@@std@@XZ) already defined in libexiv2int.a(src_cr2header_int.cpp.obj)
```

That is...very interesting

File: unitTests/meson.build [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    'test_matroskavideo.cpp',
+    'test_riffVideo.cpp',
+  ]
+endif
+
+t_args = ['-DTESTDATA_PATH="@0@"'.format('..' / 'test' / 'data')]
+
+privinc = include_directories('../src', '../include/exiv2')
+unit_tests = executable(
+  test_sources,
```
Comment (eli-schwartz) [inline unitTests/meson.build:49]: Missing executable name. This executable is therefore `unitTests/mainTestRunner.cpp.exe` and it doesn't include the main() implementation from there (instead it uses gtest_main?).

Comment (neheb) [reply unitTests/meson.build:49]: knew there was some reason files() didn't work. Thanks.

Comment (eli-schwartz) [reply unitTests/meson.build:49]: yup, in Meson git master the python traceback is fixed with real type checking:
```
unitTests/meson.build:48:0: ERROR: executable argument 1 was of type "File" but should have been "str"
```

**Comment by neheb** [comment]
fascinating. I got it to compile with MinGW but not MSVC.

**Comment by neheb** [comment]
reading http://google.github.io/googletest/primer.html#writing-the-main-function and seeing unitTests/mainTestRunner.cpp makes me thing it's completely unnecessary. Different PR of course.

@piponazo what do you think?

**Comment by neheb** [comment]
```
 stderr:
Warning: Unsupported date format
Warning: Unsupported date format
Warning: Unsupported date format
Warning: Unsupported date format
Warning: Unsupported date format
Warning: Unsupported date format
Warning: Unsupported date format
Warning: Unsupported time format
Warning: Unsupported time format
Warning: Unsupported time format
Warning: Unsupported time format
Warning: Unsupported time format
Warning: Unsupported time format
Warning: Unsupported time format
double free or corruption (!prev)
```

only on linux even though all tests passed. hmmmm.

**Comment by neheb** [comment]
uhhh fixed by using gtest-main? That can't be right...

**Comment by neheb** [comment]
issue in Linux is:

```
double free or corruption (!prev)

Program received signal SIGABRT, Aborted.
0x00007ffff72afe5c in __pthread_kill_implementation () from /lib64/libc.so.6
Missing separate debuginfos, use: dnf debuginfo-install expat-2.5.0-1.fc37.x86_64 glibc-2.36-9.fc37.x86_64 gmock-1.12.1-2.fc37.x86_64 gtest-1.12.1-2.fc37.x86_64 inih-56-2.fc37.x86_64 keyutils-libs-1.6.1-5.fc37.x86_64 krb5-libs-1.19.2-13.fc37.x86_64 libbrotli-1.0.9-9.fc37.x86_64 libcom_err-1.46.5-3.fc37.x86_64 libcurl-minimal-7.85.0-5.fc37.x86_64 libgcc-12.2.1-4.fc37.x86_64 libidn2-2.3.4-1.fc37.x86_64 libnghttp2-1.51.0-1.fc37.x86_64 libselinux-3.4-5.fc37.x86_64 libstdc++-12.2.1-4.fc37.x86_64 libunistring-1.0-2.fc37.x86_64 pcre2-10.40-1.fc37.1.x86_64 zlib-1.2.12-5.fc37.x86_64
(gdb) bt
#0  0x00007ffff72afe5c in __pthread_kill_implementation () from /lib64/libc.so.6
#1  0x00007ffff725fa76 in raise () from /lib64/libc.so.6
#2  0x00007ffff72497fc in abort () from /lib64/libc.so.6
#3  0x00007ffff72a408e in __libc_message () from /lib64/libc.so.6
#4  0x00007ffff72b9b9c in malloc_printerr () from /lib64/libc.so.6
#5  0x00007ffff72bbccc in _int_free () from /lib64/libc.so.6
#6  0x00007ffff72be363 in free () from /lib64/libc.so.6
#7  0x0000000000643edf in std::__new_allocator<std::__detail::_Hash_node_base*>::deallocate (this=0x7fffffffe177, __p=0x7f7280, __n=541)
    at /usr/include/c++/12/bits/new_allocator.h:158
#8  0x0000000000642dc6 in std::allocator_traits<std::allocator<std::__detail::_Hash_node_base*> >::deallocate (__a=..., __p=0x7f7280, __n=541)
    at /usr/include/c++/12/bits/alloc_traits.h:496
#9  0x000000000064222a in std::__detail::_Hashtable_alloc<std::allocator<std::__detail::_Hash_node<std::pair<std::pair<unsigned int, Exiv2::IfdId> const, std::unique_ptr<Exiv2::Internal::TiffComponent, std::default_delete<Exiv2::Internal::TiffComponent> > (*)(unsigned short, Exiv2::IfdId)>, true> > >::_M_deallocate_buckets (this=0x7a44a0 <Exiv2::Internal::TiffCreator::tiffGroupTable_>, __bkts=0x7f7280, __bkt_count=541)
    at /usr/include/c++/12/bits/hashtable_policy.h:2032
#10 0x0000000000640aec in std::_Hashtable<std::pair<unsigned int, Exiv2::IfdId>, std::pair<std::pair<unsigned int, Exiv2::IfdId> const, std::unique_ptr<Exiv2::Internal::TiffComponent, std::default_delete<Exiv2::Internal::TiffComponent> > (*)(unsigned short, Exiv2::IfdId)>, std::allocator<std::pair<std::pair<unsigned int, Exiv2::IfdId> const, std::unique_ptr<Exiv2::Internal::TiffComponent, std::default_delete<Exiv2::Internal::TiffComponent> > (*)(unsigned short, Exiv2::IfdId)> >, std::__detail::_Select1st, std::equal_to<std::pair<unsigned int, Exiv2::IfdId> >, Exiv2::Internal::TiffGroupKey_hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_deallocate_buckets (
    this=0x7a44a0 <Exiv2::Internal::TiffCreator::tiffGroupTable_>, __bkts=0x7f7280, __bkt_count=541) at /usr/include/c++/12/bits/hashtable.h:452
#11 0x000000000063f0c4 in std::_Hashtable<std::pair<unsigned int, Exiv2::IfdId>, std::pair<std::pair<unsigned int, Exiv2::IfdId> const, std::unique_ptr<Exiv2::Internal::TiffComponent, std::default_delete<Exiv2::Internal::TiffComponent> > (*)(unsigned short, Exiv2::IfdId)>, std::allocator<std::pair<std::pair<unsigned int, Exiv2::IfdId> const, std::unique_ptr<Exiv2::Internal::TiffComponent, std::default_delete<Exiv2::Internal::TiffComponent> > (*)(unsigned short, Exiv2::IfdId)> >, std::__detail::_Select1st, std::equal_to<std::pair<unsigned int, Exiv2::IfdId> >, Exiv2::Internal::TiffGroupKey_hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::_M_deallocate_buckets (
    this=0x7a44a0 <Exiv2::Internal::TiffCreator::tiffGroupTable_>) at /usr/include/c++/12/bits/hashtable.h:457
#12 0x000000000063dbec in std::_Hashtable<std::pair<unsigned int, Exiv2::IfdId>, std::pair<std::pair<unsigned int, Exiv2::IfdId> const, std::unique_ptr<Exiv2::Internal::TiffComponent, std::default_delete<Exiv2::Internal::TiffComponent> > (*)(unsigned short, Exiv2::IfdId)>, std::allocator<std::pair<std::pair<unsigned int, Exiv2::IfdId> const, std::unique_ptr<Exiv2::Internal::TiffComponent, std::default_delete<Exiv2::Internal::TiffComponent> > (*)(unsigned short, Exiv2::IfdId)> >, std::__detail::_Select1st, std::equal_to<std::pair<unsigned int, Exiv2::IfdId> >, Exiv2::Internal::TiffGroupKey_hash, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::~_Hashtable (
    this=0x7a44a0 <Exiv2::Internal::TiffCreator::tiffGroupTable_>, __in_chrg=<optimized out>) at /usr/include/c++/12/bits/hashtable.h:1585
#13 0x0000000000645be6 in std::unordered_map<std::pair<unsigned int, Exiv2::IfdId>, std::unique_ptr<Exiv2::Internal::TiffComponent, std::default_delete<Exiv2::Internal::TiffComponent> > (*)(unsigned short, Exiv2::IfdId), Exiv2::Internal::TiffGroupKey_hash, std::equal_to<std::pair<unsigned int, Exiv2::IfdId> >, std::allocator<std::pair<std::pair<unsigned int, Exiv2::IfdId> const, std::unique_ptr<Exiv2::Internal::TiffComponent, std::default_delete<Exiv2::Internal::TiffComponent> > (*)(unsigned short, Exiv2::IfdId)> > >::~unordered_map (this=0x7a44a0 <Exiv2::Internal::TiffCreator::tiffGroupTable_>, __in_chrg=<optimized out>)
    at /usr/include/c++/12/bits/unordered_map.h:102
#14 0x00007ffff7261b97 in __cxa_finalize () from /lib64/libc.so.6
#15 0x00007ffff7ac7ad7 in __do_global_dtors_aux () from /home/mangix/devstuff/exiv2/a/unitTests/../libexiv2.so.1
#16 0x00007fffffffe3f0 in ?? ()
#17 0x00007ffff7fcca9e in _dl_fini () at dl-fini.c:142
```

why would tiffGroupTable_ be double freed? ping @kevinbackhouse

**Comment by neheb** [comment]
Fixed by replacing unordered_map with map. But I still get

```
1/1 Unit Tests FAIL            0.10s   killed by signal 11 SIGSEGV
```

even though running the tests manually succeeds, No idea what's going on here.

**Comment by neheb** [comment]
```
 ld: warning: direct access in function 'Exiv2::Jp2Image::doWriteMetadata(Exiv2::BasicIo&)' from file 'libexiv2.a(src_jp2image.cpp.o)' to global weak symbol 'std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::endl<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&)' from file 'exiv2.p/app_actions.cpp.o' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
```

is a hint to what's wrong.

**Comment by neheb** [comment]
Alright this is as good as it will ever get honestly.

**Comment by neheb** [comment]
ping @piponazo

**Comment by kevinbackhouse** [comment]
> tiffGroupTable_

Apologies for the slow reply, I was busy recently. That's very weird: `tiffGroupTable_` is a constant global, so I don't understand how that can happen.

https://github.com/Exiv2/exiv2/blob/976dcd8e7be196daee03fd3c368f2dd6b0cf27e2/src/tiffimage_int.cpp#L1411

Maybe we should make it `static` though? It's only used in `tiffimage_int.cpp`.

**Review by kevinbackhouse — APPROVED** [review]

**Comment by neheb** [comment]
Issue was mismatched visibility settings. Ignore.

**Pull request merged**